state/indexer/sink/psql/psql.go

```diff
@@ -55,116 +56,182 @@ func (es *EventSink) DB() *sql.DB { return es.store }
 	if err != nil {
 		return err
 	}
-
-	// index end_block events
-	sqlStmt, err = indexBlockEvents(
-		sqlStmt, h.ResultEndBlock.Events, types.EventTypeEndBlock, h.Header.Height, ts, es.chainID)
-	if err != nil {
+	if err := query(tx); err != nil {
+		_ = tx.Rollback() // report the initial error, not the rollback
 		return err
 	}
+	return tx.Commit()
+}
 
-	_, err = sqlStmt.RunWith(es.store).Exec()
-	return err
+// queryWithID executes the specified SQL query with the given arguments,
+// expecting a single-row, single-column result containing an ID. If the query
+// succeeds, the ID from the result is returned.
+func queryWithID(tx *sql.Tx, query string, args ...interface{}) (uint32, error) {
+	var id uint32
+	if err := tx.QueryRow(query, args...).Scan(&id); err != nil {
+		return 0, err
+	}
+	return id, nil
 }
 
-func (es *EventSink) IndexTxEvents(txr []*abci.TxResult) error {
-	// index the tx result
-	var txid uint32
-	sqlStmtTxResult := sq.
-		Insert(TableResultTx).
-		Columns("tx_result", "created_at").
-		PlaceholderFormat(sq.Dollar).
-		RunWith(es.store).
-		Suffix("ON CONFLICT (tx_result)").
-		Suffix("DO NOTHING").
-		Suffix("RETURNING \"id\"")
-
-	sqlStmtEvents := sq.
-		Insert(TableEventTx).
-		Columns("key", "value", "height", "hash", "tx_result_id", "created_at", "chain_id").
-		PlaceholderFormat(sq.Dollar).
-		Suffix("ON CONFLICT (key,hash)").
-		Suffix("DO NOTHING")
-
-	ts := time.Now()
-	for _, tx := range txr {
-		txBz, err := proto.Marshal(tx)
-		if err != nil {
-			return err
-		}
+// insertEvents inserts a slice of events and any indexed attributes of those
+// events into the database associated with tx.
+//
+// If txID > 0, the event is attributed to the transaction with that ID;
+// otherwise it is recorded as a block event.
+func insertEvents(tx *sql.Tx, blockID, txID uint32, evts []abci.Event) error {
+	// Populate the transaction ID field iff one is defined (> 0).
+	var txIDArg interface{}
+	if txID > 0 {
+		txIDArg = txID
+	}
 
-		sqlStmtTxResult = sqlStmtTxResult.Values(txBz, ts)
+	// Add each event to the events table, and retrieve its row ID to use when
+	// adding any attributes the event provides.
+	for _, evt := range evts {
+		// Skip events with an empty type.
+		if evt.Type == "" {
+			continue
+		}
 
-		// execute sqlStmtTxResult db query and retrieve the txid
-		r, err := sqlStmtTxResult.Query()
+		eid, err := queryWithID(tx, `
+INSERT INTO `+tableEvents+` (block_id, tx_id, type) VALUES ($1, $2, $3)
+  RETURNING rowid;
+`, blockID, txIDArg, evt.Type)
 		if err != nil {
 			return err
 		}
-		defer r.Close()
 
-		if !r.Next() {
-			return nil
+		// Add any attributes flagged for indexing.
+		for _, attr := range evt.Attributes {
+			if !attr.Index {
+				continue
+			}
+			compositeKey := evt.Type + "." + attr.Key
+			if _, err := tx.Exec(`
+INSERT INTO `+tableAttributes+` (event_id, key, composite_key, value)
+  VALUES ($1, $2, $3, $4);
+`, eid, attr.Key, compositeKey, attr.Value); err != nil {
```

Comment:
see above concerns about raw string concatination and constants

Reply:
My intent was that there should be no runtime string assembly in these queries, except whatever happens under the covers in the `sql` package. The table names are all compile-time constants, for exactly that reason. I just did a read back over the queries to make sure I didn't miss a case, and I think they should all be static.